### PR TITLE
James 1804 - Handle jmap text search in ES

### PR DESCRIPTION
**Pull request by aduprat**
Here is the specification from the text property on FilterCondition: "text: String|null Looks for the text in the from, to, cc, bcc, subject, textBody or htmlBody properties of the message."

The easiest way (without changing the ES indices) is to handle multiple fields in IndexableMessage. For example, IndexableMessage may have a text attibute which will be the concatenation of from, to, cc, bcc, subject, textBody and htmlBody

We may also add a body attribute for textBody and htmlBody

For information: http://stackoverflow.com/questions/22617674/does-elasticsearch-have-compound-indexes#22657490 NOTE: add a CAPABILITY

### CODE REVIEW

**Comment by aduprat** [comment]
Real commit order:
JAMES-1804 Add Text capability to MessageSearchIndex
JAMES-1804 Index htmlBody attribute
JAMES-1804 Add log when indexing
JAMES-1804 Add logback in MPT tests with DEBUG level on James
JAMES-1804 Index text attribute Append of: from, to, cc, bcc, subjects, body
JAMES-1804 Introduce type for Set<String> (Subjects)
JAMES-1804 Introduce type for Set<EMailer>

File: server/container/guice/memory-guice/src/main/java/org/apache/james/modules/mailbox/MemoryMessageSearchIndex.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+package org.apache.james.modules.mailbox;
+
+import java.util.EnumSet;
+
+import javax.inject.Inject;
+
+import org.apache.james.mailbox.store.mail.MessageMapperFactory;
+import org.apache.james.mailbox.store.search.SimpleMessageSearchIndex;
+
+public class MemoryMessageSearchIndex extends SimpleMessageSearchIndex {
```
Comment (aduprat) [inline server/container/guice/memory-guice/src/main/java/org/apache/james/modules/mailbox/MemoryMessageSearchIndex.java]: Should we plug ES in memory tests?

File: mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/json/IndexableMessage.java [hunk 2]
```diff
@@ -98,6 +101,17 @@ private static ZonedDateTime getSanitizedInternalDate(MailboxMessage message, Zo
             zoneId);
     }
 
+    private void generateText() {
+        this.text = Stream.of(from.serialize(),
+                to.serialize(),
+                cc.serialize(),
+                bcc.serialize(),
+                subjects.serialize(),
+                bodyText.orElse(null))
```
Comment (aduprat) [inline mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/json/IndexableMessage.java]: Will add bodyHtml...

File: mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/events/ElasticSearchListeningMessageSearchIndex.java [hunk 3]
```diff
@@ -62,6 +64,11 @@ public ElasticSearchListeningMessageSearchIndex(MessageMapperFactory factory, El
     }
 
     @Override
+    public boolean hasCapability(MessageSerachIndexCapabilities capability) {
```
Comment (rouazana) [inline mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/events/ElasticSearchListeningMessageSearchIndex.java]: s/Serach/Search/

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/search/MessageSearchIndex.java [hunk 4]
```diff
@@ -36,7 +36,14 @@
  * @param <Id>
  */
 public interface MessageSearchIndex {
-    
+
+
+    enum MessageSerachIndexCapabilities {
+        Text
```
Comment (rouazana) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/search/MessageSearchIndex.java]: Not trivial to guess what means "text" in a searchIndex context. Please rename it or comment it.

File: mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/json/IndexableMessage.java [hunk 5]
```diff
@@ -98,6 +101,17 @@ private static ZonedDateTime getSanitizedInternalDate(MailboxMessage message, Zo
             zoneId);
     }
 
+    private void generateText() {
+        this.text = Stream.of(from.serialize(),
```
Comment (mbaechler) [inline mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/json/IndexableMessage.java]: I would have put this code in getText() to ensure we never forget to build the field

File: mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/NodeMappingFactory.java [hunk 6]
```diff
@@ -208,14 +210,36 @@ private static XContentBuilder getMappingContent() {
...
+                                .endObject()
+                            .endObject()
 
                             .startObject(JsonMessageConstants.TEXT)
                                 .field(TYPE, STRING)
                                 .startObject(FIELDS)
                                     .startObject(RAW)
                                         .field(TYPE, STRING)
                                         .field(ANALYZER, IndexCreationFactory.CASE_INSENSITIVE)
+                                        .field(IGNORE_ABOVE, LUCENE_LIMIT)
```
Comment (mbaechler) [inline mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/NodeMappingFactory.java]: what does this mean ?

**Comment by mbaechler** [comment]
You really miss some tests on the new indices.

File: mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/events/ElasticSearchListeningMessageSearchIndex.java (hunk 3 above)
Comment (aduprat) [inline mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/events/ElasticSearchListeningMessageSearchIndex.java]: OK

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/search/MessageSearchIndex.java (hunk 4 above)
Comment (aduprat) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/search/MessageSearchIndex.java]: OK

File: mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/json/IndexableMessage.java (hunk 5 above)
Comment (aduprat) [inline mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/json/IndexableMessage.java]: I don't think, as
* the only valid way to instanciate an IndexableMessage is to call the static method `from`
* joining large strings may have a cost
* there are no setter on the IndexableMessage attributes

File: mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/NodeMappingFactory.java (hunk 6 above)
Comment (aduprat) [inline mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/NodeMappingFactory.java]: While coding unit-test, i have encountered this issue; which is a Lucene limitation:
http://stackoverflow.com/questions/24019868/utf8-encoding-is-longer-than-the-max-length-32766

**Comment by aduprat** [comment]
Tests added

**Comment by rouazana** [comment]
:+1:

**Comment by aduprat** [comment]
test this please

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/search/AbstractMessageSearchIndexTest.java [hunk 7]
```diff
@@ -720,4 +720,22 @@ public void searchWithFullTextShouldReturnMailsWhenTextBodyMatchesAndNonContinuo
...
+    public void searchWithFullTextShouldReturnMailsWhenHtmlBodyMatches() throws Exception {
+        Assume.assumeTrue(messageSearchIndex.hasCapability(MessageSearchIndexCapabilities.Text));
+        SearchQuery searchQuery = new SearchQuery();
+        searchQuery.andCriteria(SearchQuery.textContains("Regarder"));
+        assertThat(messageSearchIndex.search(session, mailbox, searchQuery))
+            .containsExactly(7l);
+    }
+
+    @Test
+    public void searchWithFullTextShouldReturnMailsWhenHtmlBodyMatchesAndNonContinuousWords() throws Exception {
```
Comment (mbaechler) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/search/AbstractMessageSearchIndexTest.java]: you miss the case-insensitive test I guess

Comment (mbaechler) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/search/AbstractMessageSearchIndexTest.java]: and tests with :
- some 'a', 'the', 'of'
- matching like "regardons" for word "regarder"

**Pull request closed without merging**